packages/dds/tree/src/codec/codec.ts

```diff
@@ -0,0 +1,177 @@
+ * Licensed under the MIT License.
+ */
+
+import { assert, bufferToString, IsoBuffer } from "@fluidframework/common-utils";
+import { fail, JsonCompatibleReadOnly } from "../util";
+
+/**
+ * @alpha
+ */
+export interface IEncoder<TDecoded, TEncoded> {
+	/**
+	 * Encodes `obj` into some encoded format. Typically paired with an {@link IDecoder}.
+	 */
+	encode(obj: TDecoded): TEncoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IDecoder<TDecoded, TEncoded> {
+	/**
+	 * Decodes `obj` from some encoded format. Typically paired with an {@link IEncoder}.
+	 */
+	decode(obj: TEncoded): TDecoded;
+}
+
+/**
+ * @alpha
+ */
+export interface IJsonCodec<TDecoded, TEncoded = JsonCompatibleReadOnly>
+	extends IEncoder<TDecoded, TEncoded>,
+		IDecoder<TDecoded, TEncoded> {}
+
+/**
+ * @remarks - TODO: We might consider using DataView or some kind of writer instead of IsoBuffer.
+ * @alpha
+ */
+export interface IBinaryCodec<TDecoded>
+	extends IEncoder<TDecoded, IsoBuffer>,
+		IDecoder<TDecoded, IsoBuffer> {}
+
+/**
+ * Contains knowledge of how to encode some in-memory type into JSON and binary formats,
+ * as well as how to decode those representations.
+ *
+ * @remarks - Codecs are typically used in shared-tree to convert data into some persisted format.
+ * For this common use case, any format for encoding that was ever actually used needs to
+ * be supported for decoding in all future code versions.
+ *
+ * Using an {@link ICodecFamily} is the recommended strategy for managing this support, keeping in
+ * mind evolution of encodings over time.
+ *
+ * @alpha
+ */
+export interface IMultiFormatCodec<TDecoded, TJsonEncoded = JsonCompatibleReadOnly> {
+	json: IJsonCodec<TDecoded, TJsonEncoded>;
+	binary: IBinaryCodec<TDecoded>;
+}
+
+/**
+ * Represents a family of codecs that can be used to encode and decode data in different formats.
+ * The family is identified by a format version, which is typically used to select the codec to use.
+ *
+ * Separating codecs into families rather than having a single codec support multiple versions (i.e. currying
+ * the `formatVersion` parameter)
+ * allows avoiding some duplicate work at encode/decode time, since the vast majority of document usage will not
+ * involve mixed format versions.
+ * @alpha
+ */
+export interface ICodecFamily<TDecoded> {
+	/**
+	 * @returns - a codec that can be used to encode and decode data in the specified format.
+	 */
+	resolve(formatVersion: number): IMultiFormatCodec<TDecoded>;
+
+	/**
+	 * @returns - an iterable of all format versions supported by this family.
+	 */
+	getSupportedFormats(): Iterable<number>;
+}
+
+/**
+ * Creates a codec family from a registry of codecs.
+ * Any codec that is not a {@link IMultiFormatCodec} will be wrapped with a default binary encoding.
+ */
+export function makeCodecFamily<TDecoded>(
+	registry: Iterable<
+		[formatVersion: number, codec: IMultiFormatCodec<TDecoded> | IJsonCodec<TDecoded>]
+	>,
+): ICodecFamily<TDecoded> {
+	const isJsonCodec = (
```

Comment:
This check isn't fully safe since you can implement IMultiFormatCodec and also have encode and decode functions.

Might make more sense for this to take a: `{ json: ..., binary?: ...} | { json?: ..., binary: ...}`. This pattern can't result in runtime vs compile time confusion as easily and avoids the need for custom type narrowing. It also generalizes to supporting more formats easily (like you could add a text format which defaults to searlizing the json or base64ing the binary)

Reply:
i'd rather make it convenient to specify codecs. I had this function taking in only a `IMultiFormatCodec` initially and it wasn't very ergonomic in calling code.

I pushed an alternative way to make it typesafe (explicitly disallow encode/decode on IMultiFormatCodec).

```typescript
export interface IMultiFormatCodec<TDecoded, TJsonEncoded = JsonCompatibleReadOnly> {
	json: IJsonCodec<TDecoded, TJsonEncoded>;
	binary: IBinaryCodec<TDecoded>;

	encode?: never;
	decode?: never;
}
```

which eliminates the need for `isJsonCodec` just as well.

Allowing a multi-format codec with some properties optional and injecting adapters for unspecified ones seems interesting, but I think we should cross that bridge when we come to it: we already only have a single codec which defines a (trivial) binary format. This API isn't exported anyway by the package, so we can change it as we deem appropriate.

Reply:
well, at least newer versions of TS make the type comparison function unnecessary. Current one doesn't do narrowing well enough.